dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/validator/relationship/RelationshipValidatorTest.java

```diff
@@ -0,0 +1,191 @@
+
+        assertEquals( 1, reporter.getErrors().size(),
+            "Expected just one error when uid not correct, but " + reporter.getErrors().size() + " found" );
+        assertHasError( reporter, relationship, E1048, "uid: `wrong uid`, has an invalid uid format." );
+    }
+
+    @Test
+    void shouldFailWhenRelationshipFromNotSet()
+    {
+        Relationship relationship = new Relationship();
+        relationship.setRelationship( CodeGenerator.generateUid() );
+        relationship.setTo( new RelationshipItem( "entity", "", "" ) );
+        relationship.setRelationshipType( MetadataIdentifier.ofUid( "uid" ) );
+        when( bundle.getRelationships() ).thenReturn( List.of( relationship ) );
+        when( bundle.getPreheat() ).thenReturn( preheat );
+
+        validator.validate( reporter, bundle, bundle );
+
+        assertEquals( 1, reporter.getErrors().size(),
+            "Expected just one error when from constraint not set, but " + reporter.getErrors().size() + " found" );
+        assertHasError( reporter, relationship, E1124, "Missing required relationship property: `from`." );
+    }
+
+    @Test
+    void shouldFailWhenRelationshipFromItemTypeNotSet()
+    {
+        String uid = CodeGenerator.generateUid();
+        Relationship relationship = createProgramRelationship( "", uid );
+        when( bundle.getRelationships() ).thenReturn( List.of( relationship ) );
+        when( preheat.getRelationshipType( relationship.getRelationshipType() ) )
+            .thenReturn( createTrackedEntityRelationship() );
+        when( bundle.getPreheat() ).thenReturn( preheat );
+
+        validator.validate( reporter, bundle, bundle );
+
+        assertEquals( 1, reporter.getErrors().size(),
+            "Expected just one error when from item type not set, but " + reporter.getErrors().size() + " found" );
+        assertHasError( reporter, relationship, E4001, "Relationship Item `from` for Relationship `" + uid
+            + "` is invalid: an Item can link only one Tracker entity." );
+    }
+
+    @Test
+    void shouldFailWhenRelationshipLinksToItself()
+    {
+        String uid = CodeGenerator.generateUid();
+        Relationship relationship = createProgramRelationship( "enrollment", uid );
+        when( bundle.getRelationships() ).thenReturn( List.of( relationship ) );
+        when( preheat.getRelationshipType( relationship.getRelationshipType() ) )
+            .thenReturn( createTrackedEntityRelationship() );
+        when( bundle.getPreheat() ).thenReturn( preheat );
+
+        validator.validate( reporter, bundle, bundle );
+
+        assertEquals( 1, reporter.getErrors().size(), "Expected just one error when relationship links to itself, but "
+            + reporter.getErrors().size() + " found" );
+        assertHasError( reporter, relationship, E4000, "Relationship: `" + uid + "` cannot link to itself" );
+    }
+
+    @Test
+    void shouldFailWhenRelationshipLinksProgramsAndRelationshipTypeIsNotSet()
```

Comment:
This test name seems wrong to me. RelationType is set; the validation fails because the items are TEIs instead of enrollments